12-factor docs: How to add a custom action (#2256)
canonical/charmcraft: docs/howto/manage-web-app-charms/configure-web-app-charm.rst

    :language: yaml
    :caption: charmcraft.yaml

Modify the ``src/charm.py`` to define your custom action as part of the class
and provide the function definition:

.. literalinclude:: ../code/custom-action-clearsession/clearsession_action_charm.py
    :language: python
    :caption: src/charm.py
    :lines: 20-46
    :emphasize-lines: 8-27

Build the charm using ``charmcraft pack`` and deploy the app with Juju.

Finally, call the action using:

.. code-block:: bash

    juju run <django unit name> clearsession

If successful, the terminal will output something like:

.. terminal::
    :input: juju run django-app/0 clearsession

    Running operation 1 with 1 task
      - task 2 on unit-django-app-0

    Waiting for task 2...
    result: session cleared!

Example: check Flask app status and write output to container file
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

As another example, let's say you want to add a custom action to your charmed
Flask app that checks the status of the running app and writes the status to a
log file in the app container. The action performs the following steps:

- Sends a request to the Flask app at its available port.

> - If the request is successful, updates the event container with the status

Reviewer: ```suggestion
- If the request is successful, updates the app container with the status
```

  message in a logfile.

> - Reads the contents of the logfile in the event container and outputs the
>   status message as part of the action result.

Reviewer: I think that someone reading this (and the code) might wonder why we push the file to the container then immediately pull it back again. How about adding a couple of extra parts:

```suggestion
- As a consistency check, reads the contents of the log file in the app
  container and outputs the status message as part of the action result.
  In practice, you could create a separate action for reading the logfile.
```

    juju config <app name> token=<token>

Add a custom action
-------------------

> For certain commands or processes that you will run many times, you can set up
> custom Juju actions to expedite your workflow. Actions are handled in the Ops

Reviewer 2: > you will run many times

The "you will" throws me a bit here because of the tone of a how-to. Can we be more descriptive with:

```suggestion
For commands and processes that that see shared or frequent use,
custom Juju actions can expedite your workflow. Actions are handled in the Ops
```

library, and Juju manages the action's workflow via tasks and operations. To
implement a custom action, you need to declare the action in

> ``charmcraft.yaml`` and define an event handler into the Ops framework of your

Reviewer 2: > define... into

"into" wouldn't be my choice for a preposition. Should it be "for"? Or maybe "add... to"?

charm.

.. seealso::

    :external+ops:ref:`Ops | Manage actions <manage-actions>`

> The custom action must be defined in ``charmcraft.yaml`` under the

Reviewer 2: ```suggestion
The custom action must be defined in the project file under the
```

I suppose it's time to bring this phrase up. It's the generic phrase for all `*craft.yaml` files in our products, and reduces the amount of code in the text.

Kindly consider it for all instances.

``actions`` section. You should provide a name, a description, and any
associated parameters.

Add the custom action to your charm code by modifying the ``src/charm.py``
file. You need to add the action under the class constructor function
(``__init__``) as well as a dedicated function for the action. When you define
the action, follow the convention of appending ``_action`` onto the
action name.

Once your web app is updated and deployed with Juju, you can call the custom
action using ``juju run``.

Example: clear cookies in a Django app
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

For example, let's say you want to add a custom action to your charmed Django
app that removes expired session cookies.

Add the custom action in ``charmcraft.yaml``:

.. literalinclude:: ../code/custom-action-clearsession/clearsession_action_charmcraft.yaml

>     :language: yaml

Reviewer 2: ```suggestion
    :language: yaml
    :caption: charmcraft.yaml
```

It would be nice if all the code blocks had these.

Modify the ``src/charm.py`` to define your custom action as part of the class
and provide the function definition:

.. literalinclude:: ../code/custom-action-clearsession/clearsession_action_charm.py
    :language: python
    :lines: 20-46
    :emphasize-lines: 8-27

Build the charm using ``charmcraft pack`` and deploy the app with Juju.

> Finally, call the action using:

Reviewer 2: ```suggestion

Finally, call the action:
```

.. code-block:: bash

    juju run <django unit name> clearsession

If successful, the terminal will output something like:

.. terminal::
    :input: juju run django-app/0 clearsession

    Running operation 1 with 1 task
      - task 2 on unit-django-app-0

    Waiting for task 2...
    result: session cleared!

Example: check Flask app status and write output to container file
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

As another example, let's say you want to add a custom action to your charmed
Flask app that checks the status of the running app and writes the status to a

> logfile in the app container. The action performs the following steps:

Reviewer 2: ```suggestion
log file in the app container. The action performs the following steps:
```

- Sends a request to the Flask app at its available port.
- If the request is successful, updates the event container with the status
  message in a logfile.

> Add the custom action in ``charmcraft.yaml``:

Reviewer 2: ```suggestion
Add the custom action to ``charmcraft.yaml``:
```

.. literalinclude:: ../code/custom-action-updatelogfile/updatelogfile_action_charmcraft.yaml
    :language: yaml

> Modify the ``src/charm.py`` to include ``import requests`` at the top, define

Reviewer 2: ```suggestion
Add ``import requests`` to the start of ``src/charm.py``, then define
```

your custom action as part of the class and provide the function definition:

.. literalinclude:: ../code/custom-action-updatelogfile/updatelogfile_action_charm.py
    :language: python
    :lines: 22-54
    :emphasize-lines: 8-33

Build the charm using ``charmcraft pack`` and deploy the app with Juju.
Finally, call the action using:

.. code:: bash

   juju run <flask unit name> updatelogfile logfile=<full path to logfile>

If successful, the terminal will output something like:

.. terminal::
    :input: juju run flask-app/0 updatelogfile logfile="/tmp/example.log"

    Running operation 1 with 1 task
      - task 2 on unit-flask-app-0

    Waiting for task 2...
    result: |-
      Hello, world!
       written to file /tmp/example.log in app container

> Check that the file was updated using Pebble:

Reviewer 2: This feels like the step that ought to start with "finally".
